### PR TITLE
[Compatibility] Add timeout argument to Thread::SizedQueue#pop

### PR DESCRIPTION
**Pull request by itarato**
Source: https://github.com/oracle/truffleruby/issues/3039

Thread::SizedQueue#pop(timeout: sec) is added. [[Feature #18774](https://bugs.ruby-lang.org/issues/18774)]

### CODE REVIEW

File: src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java [hunk 1]
```diff
@@ -182,8 +178,36 @@ private Object doPop(SizedQueue queue) {
...
+                } else {
+                    value = queue.poll();
+                }
+
+                if (value == SizedQueue.CLOSED) {
+                    throw new RaiseException(getContext(), coreExceptions().closedQueueError(this));
+                } else {
+                    return value;
+                }
+            });
```
Comment (andrykonchin) [inline src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java:201]: Wondering how a case with expired timeout is handled. Could we debug this case?

When a blocking `pop` call with timeout happens and timeout expires then I suppose the following things happen (please correct me if I am wrong):
1. `canTake.await` returns `false` in `SizedQueue.poll`
2. `SizedQueue.poll` returns `null`
3. `null` is returned from a lambda, passed as argument to the `runUntilResult` method
4. `runUntilResult` re-runs the lambda as far as it's waiting for the first non-`null` result
5. as far as timeout is expired then `currentTimeout > 0` is `false` and `queue.poll()` is called
6. `queue.poll()` returns `null` for an empty queue (our queue is still empty as far as `poll` time outed)
7. and `runUntilResult` re-runs the lambda again and again until `queue.poll()`  returns non-`null` value

So this way a blocking `pop` call with timeout when a queue is empty will wait longer than the specified timeout -  until some value is added to the queue.

File: CHANGELOG.md [hunk 2]
```diff
@@ -22,6 +22,7 @@ Compatibility:
 * Add `String#byteindex` and `String#byterindex` (#3039, @itarato).
 * Add implementations of `rb_proc_call_with_block`, `rb_proc_call_kw`, `rb_proc_call_with_block_kw` and `rb_funcall_with_block_kw` (#3068, @andrykonchin).
 * Add optional `timeout` argument to `Thread::Queue#pop` (#3039, @itarato).
+* Add optional `timeout` argument to `Thread::SizedsQueue#pop` (@itarato).
```
[the remark this answers is not in the document]
Comment (itarato) [reply CHANGELOG.md:25]: Done

File: src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java [hunk 3]
```diff
@@ -182,8 +178,36 @@ private Object doPop(SizedQueue queue) {
...
+
+            if (result == null) {
+                return nil;
+            }
+            return result;
+        }
+
         @Specialization(guards = "nonBlocking")
-        protected Object popNonBlock(RubySizedQueue self, boolean nonBlocking,
+        protected Object popNonBlock(RubySizedQueue self, boolean nonBlocking, Nil timeoutMilliseconds,
```
[the remark this answers is not in the document]
Comment (itarato) [reply src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java:210]: That case is guarded in ruby: https://github.com/oracle/truffleruby/blob/c6d73dd6f04c6420b7b9833e27b00bffab87be91/src/main/ruby/truffleruby/core/truffle/queue_operations.rb#L15

File: lib/truffle/thread.rb [hunk 4]
```diff
@@ -40,4 +40,15 @@ def pop(non_block = false, timeout: nil)
     alias_method :shift, :pop
     alias_method :deq, :pop
   end
+
+  class SizedQueue
+    def pop(non_block = false, timeout: nil)
+      Primitive.sized_queue_pop(
+        self,
+        Primitive.as_boolean(non_block),
+        Truffle::QueueOperations.validate_and_prepare_timeout(non_block, timeout))
```
[the remark this answers is not in the document]
Comment (itarato) [reply lib/truffle/thread.rb:49]: Agree, updating

File: src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrykonchin) [reply src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java:210]: Right 👍

File: src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java [hunk 5]
```diff
@@ -182,8 +178,36 @@ private Object doPop(SizedQueue queue) {
...
+                final Object value;
+
+                if (currentTimeout > 0) {
+                    value = queue.poll(currentTimeout);
+                } else {
+                    value = queue.poll();
+                }
+
+                if (value == SizedQueue.CLOSED) {
+                    throw new RaiseException(getContext(), coreExceptions().closedQueueError(this));
```
[the remark this answers is not in the document]
Comment (itarato) [reply src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java:197]: Yes, my bad, fixing it

File: lib/truffle/thread.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrykonchin) [reply lib/truffle/thread.rb:49]: **subjective:** It seems to me that this small helper (I mean `validate_and_prepare_timeout`) could be refactored and simplified a bit with guard conditions

```ruby
if non_block
  raise ArgumentError, "can't set a timeout if non_block is enabled" unless Primitive.nil?(timeout)
else
  unless Primitive.nil?(timeout)
    # Convert seconds to milliseconds.
    return (Truffle::Type.rb_num2dbl(timeout) * 1000).to_i
  end
end

timeout
```

e.g. in the following way (the comment is removed as far as the helper name reflects the logic):

```ruby
return nil if Primitive.nil?(timeout)
raise ArgumentError, "can't set a timeout if non_block is enabled" if non_block
(Truffle::Type.rb_num2dbl(timeout) * 1000).to_i
```

File: src/main/java/org/truffleruby/core/queue/SizedQueue.java [hunk 6]
```diff
@@ -137,6 +138,33 @@ private void doAdd(Object item) {
...
+            if (size == 0) {
+                final boolean signalled = canTake.await(timeoutMilliseconds, TimeUnit.MILLISECONDS);
+
+                if (Thread.interrupted()) {
+                    throw new InterruptedException();
+                }
+
+                if (!signalled) {
+                    return null;
+                }
```
[the remark this answers is not in the document]
Comment (itarato) [reply src/main/java/org/truffleruby/core/queue/SizedQueue.java:155]: Not sure I follow. Reading `Condition#await(t, unit)` docs (which gives `signalled`):

```
     * @return {@code false} if the waiting time detectably elapsed
     *         before return from the method, else {@code true}
```

It seems it's the basic knowledge whether it has timed out or not - or put it simply, we return `null` when there is still nothing to pop after the timeout. Maybe I'm not seeing something deeper here.

File: lib/truffle/thread.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (itarato) [reply lib/truffle/thread.rb]: I like it

File: src/main/java/org/truffleruby/core/queue/SizedQueue.java [hunk 7]
```diff
@@ -137,6 +138,33 @@ private void doAdd(Object item) {
...
+    public Object poll(long timeoutMilliseconds) throws InterruptedException {
+        lock.lock();
+
+        try {
+            if (size == 0) {
+                final boolean signalled = canTake.await(timeoutMilliseconds, TimeUnit.MILLISECONDS);
+
+                if (Thread.interrupted()) {
+                    throw new InterruptedException();
+                }
```
[the remark this answers is not in the document]
Comment (andrykonchin) [reply src/main/java/org/truffleruby/core/queue/SizedQueue.java:151]: Got it. So it makes sense to have the same comment here as well.

File: src/main/java/org/truffleruby/core/queue/SizedQueue.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrykonchin) [reply src/main/java/org/truffleruby/core/queue/SizedQueue.java:155]: Ah, I just have figured out that `Condition#await` releases a connected `lock` so elements can be pushed into a queue.

File: src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java (hunk 1 above)
Comment (andrykonchin) [reply src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java:201]: My bad. It seems `runUntilResult` doesn't wait for the first non-null value.

**Review by andrykonchin — APPROVED** [review]

File: src/main/java/org/truffleruby/core/queue/SizedQueue.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (itarato) [reply src/main/java/org/truffleruby/core/queue/SizedQueue.java:155]: Sure.

File: src/main/java/org/truffleruby/core/queue/SizedQueue.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (itarato) [reply src/main/java/org/truffleruby/core/queue/SizedQueue.java:151]: Indeed.

**Review by eregon — APPROVED** [review]

**Pull request merged**